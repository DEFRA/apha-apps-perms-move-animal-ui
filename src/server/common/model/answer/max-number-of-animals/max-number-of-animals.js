import { NumberAnswer } from '../number/number.js'

const min = 1
const max = 5000

/** @import { NumberConfig } from '../number/number.js' */

/**
 * export @typedef {string} MaxNumberOfAnimalsData
 * @typedef {{ maxNumberOfAnimals: string }} MaxNumberOfAnimalsPayload
 */

/**
 * @augments {NumberAnswer<MaxNumberOfAnimalsPayload>}
 */
export class MaxNumberOfAnimalsAnswer extends NumberAnswer {
  /** @type {NumberConfig} */
  static config = {
    payloadKey: 'maxNumberOfAnimals',
    autocomplete: 'maxNumberOfAnimals',
    characterWidth: 2,
    validation: {
      empty: { message: 'Enter how many animals you are planning to move' },
      max: { value: max, message: `Enter a number ${max} or below` },
      min: { value: min, message: `Enter a number ${min} or above` }
    }
  }
}
