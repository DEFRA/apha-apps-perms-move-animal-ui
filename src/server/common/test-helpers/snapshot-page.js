import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
