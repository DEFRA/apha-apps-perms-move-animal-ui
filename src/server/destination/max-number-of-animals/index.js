import { QuestionPageController } from '../../common/controller/question-page-controller/question-page-controller.js'
import { HowManyAnimalsAnswer } from '../../common/model/answer/max-number-of-animals/max-number-of-animals.js'
import { QuestionPage } from '../../common/model/page/question-page-model.js'
import { reasonForMovementPage } from '../reason-for-movement/index.js'

export class HowManyAnimalsPage extends QuestionPage {
  urlPath = '/destination/how-many-animals'
  sectionKey = 'destination'
  question = 'How many animals are you planning to move?'
  questionKey = 'howManyAnimals'

  Answer = HowManyAnimalsAnswer

  nextPage() {
    return reasonForMovementPage
  }
}

export const howManyAnimalsPage = new HowManyAnimalsPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const howManyAnimals = new QuestionPageController(
  howManyAnimalsPage
).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
