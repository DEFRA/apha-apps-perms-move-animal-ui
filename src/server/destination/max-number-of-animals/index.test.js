import { maxNumberOfAnimalsPage, MaxNumberOfAnimalsPage } from './index.js'
import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'
import { MaxNumberOfAnimalsAnswer } from '../../common/model/answer/max-number-of-animals/max-number-of-animals.js'
import { reasonForMovementPage } from '../reason-for-movement/index.js'

const sectionKey = 'destination'
const question = 'Enter the maximum number of animals you are planning to move'
const questionKey = 'maxNumberOfAnimals'
const view = 'common/model/page/question-page.njk'
const pageUrl = '/destination/how-many-animals'

describe('MaxNumberOfAnimalsPage', () => {
  let page

  beforeEach(() => {
    page = new MaxNumberOfAnimalsPage()
  })

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(MaxNumberOfAnimalsAnswer)
  })

  it('nextPage should return address page', () => {
    const nextPage = page.nextPage()
    expect(nextPage).toBe(reasonForMovementPage)
  })

  it('should export page', () => {
    expect(maxNumberOfAnimalsPage).toBeInstanceOf(MaxNumberOfAnimalsPage)
  })

  describePageSnapshot({
    describes: 'MaxNumberOfAnimalsPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
