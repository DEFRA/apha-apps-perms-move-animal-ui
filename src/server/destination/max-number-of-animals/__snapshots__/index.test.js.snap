// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxNumberOfAnimalsPage MaxNumberOfAnimalsPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<fieldset class="govuk-fieldset">
  <div class="govuk-form-group">
  <h1 class="govuk-label-wrapper">
    <label class="govuk-label govuk-label--l" for="maxNumberOfAnimals">
      Enter the maximum number of animals you are planning to move
    </label>
  </h1>
  <input class="govuk-input govuk-input--width-2" id="maxNumberOfAnimals" name="maxNumberOfAnimals" type="text" autocomplete="maxNumberOfAnimals">
</div>


</fieldset>
            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
